dali/operators/image/resize/resize.cc

```diff
@@ -70,61 +25,81 @@ DALI_SCHEMA(Resize)
+  InitializeBackend();
 }
 
 template <>
-void Resize<CPUBackend>::SetupSharedSampleParams(SampleWorkspace &ws) {
-  const int thread_idx = ws.thread_idx();
-  per_sample_meta_[thread_idx] = GetTransfomMeta(&ws, spec_);
-  resample_params_[thread_idx] = detail::GetResamplingParams(
-    per_sample_meta_[thread_idx], min_filter_, mag_filter_);
+void Resize<CPUBackend>::InitializeBackend() {
+  InitializeCPU(num_threads_);
 }
 
 template <>
-void Resize<CPUBackend>::RunImpl(SampleWorkspace &ws) {
+void Resize<CPUBackend>::RunImpl(HostWorkspace &ws) {
   const int thread_idx = ws.thread_idx();
-  const auto &input = ws.Input<CPUBackend>(0);
-  auto &output = ws.Output<CPUBackend>(0);
-
-  DALI_ENFORCE(IsType<uint8>(input.type()), "Expected input data as uint8.");
-  DALI_ENFORCE(input.ndim() == 3, "Resize expects 3-dimensional tensor input.");
-  if (!input.GetLayout().empty()) {
-    DALI_ENFORCE(ImageLayoutInfo::IsChannelLast(input.GetLayout()),
-                 "Resize expects interleaved channel layout aka (N)HWC");
-  }
+  const auto &input = ws.InputRef<CPUBackend>(0);
+  auto &output = ws.OutputRef<CPUBackend>(0);
 
-  RunCPU(output, input, thread_idx);
+  RunResize(ws, output, input);
   output.SetLayout(input.GetLayout());
 
   if (save_attrs_) {
-    auto &attr_output = ws.Output<CPUBackend>(1);
-    auto &in_shape = input.shape();
-
-    attr_output.Resize({2});
-    int *t = attr_output.mutable_data<int>();
-    t[0] = in_shape[0];
-    t[1] = in_shape[1];
+    const auto &input_shape = input.shape();
+    auto &attr_out = ws.OutputRef<CPUBackend>(1);
+    const auto &attr_shape = attr_out.shape();
+    assert(attr_shape.num_samples() == input_shape.num_samples() &&
+          attr_shape.sample_dim() == 1 &&
+          is_uniform(attr_shape) &&attr_shape[0][0] == NumSpatialDims());
+
+    auto attr_view = view<int, 1>(attr_out);
+    SaveAttrs(attr_view, input.shape());
   }
 }
 
 DALI_REGISTER_OPERATOR(Resize, Resize<CPUBackend>, CPU);
 
+
+template <>
+void Resize<GPUBackend>::InitializeBackend() {
+  InitializeGPU(spec_.GetArgument<int>("minibatch_size"));
+}
+
+template<>
+void Resize<GPUBackend>::RunImpl(DeviceWorkspace &ws) {
+  const auto &input = ws.Input<GPUBackend>(0);
+  auto &output = ws.Output<GPUBackend>(0);
+
+  RunResize(ws, output, input);
+  output.SetLayout(input.GetLayout());
+
+  if (save_attrs_) {
+    auto &attr_out = ws.Output<GPUBackend>(1);
+    const auto &attr_shape = attr_out.shape();
+    assert(attr_shape.num_samples() == input.shape().num_samples() &&
+           attr_shape.sample_dim() == 1 &&
+           is_uniform(attr_shape) &&
+           attr_shape[0][0] == NumSpatialDims());
+
+    if (!attr_staging_.raw_data())
+      attr_staging_.set_pinned(true);
+    attr_staging_.ResizeLike(attr_out);
+    auto attr_view = view<int, 1>(attr_staging_);
+    SaveAttrs(attr_view, input.shape());
+    attr_out.Copy(attr_staging_, ws.stream());
+  }
+}
+
+DALI_REGISTER_OPERATOR(Resize, Resize<GPUBackend>, GPU);
```

Comment:
Shouldn't be this moved to *.cu file?

Reply:
I'm not using any CUDA code - it would just take longer to compile and give poor diagnostics.